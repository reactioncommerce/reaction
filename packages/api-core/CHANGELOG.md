# @reactioncommerce/api-core

## 3.0.0

### Major Changes

- [#6740](https://github.com/reactioncommerce/reaction/pull/6740) [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9) Thanks [@sujithvn](https://github.com/sujithvn)! - Node18 upgrade includes assert json

### Minor Changes

- [#6684](https://github.com/reactioncommerce/reaction/pull/6684) [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be) Thanks [@sujithvn](https://github.com/sujithvn)! - Filter feature. This new feature provides a common function that can be used in a new query endpoint to get filtered results from any collection.

- [#6739](https://github.com/reactioncommerce/reaction/pull/6739) [`0b4b3385f`](https://github.com/reactioncommerce/reaction/commit/0b4b3385f218c4085d8d9f641a28a6e567f4c918) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to verion 5.0

- [#6588](https://github.com/reactioncommerce/reaction/pull/6588) [`9d26f0528`](https://github.com/reactioncommerce/reaction/commit/9d26f05286a43d99f4b8a2924b5a0644a5a03ccc) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to 4.4

### Patch Changes

- Updated dependencies [[`a50a7b359`](https://github.com/reactioncommerce/reaction/commit/a50a7b359bbb546b7abab0e0bfed4c5d8b5ad759), [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be), [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9)]:
  - @reactioncommerce/api-utils@2.0.0
