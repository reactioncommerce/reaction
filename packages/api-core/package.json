{
  "name": "@reactioncommerce/api-core",
  "version": "3.0.0",
  "description": "This NPM package provides the `ReactionAPICore` class. Use this to build a NodeJS microservice that is compatible with the Reaction Commerce platform, or to build your main Reaction Commerce API if you don't want to start by forking the `https://github.com/reactioncommerce/reaction` project.",
  "main": "./src/index.js",
  "type": "module",
  "sideEffects": false,
  "engines": {
    "node": ">=14.18.1"
  },
  "homepage": "https://github.com/reactioncommerce/reaction",
  "url": "https://github.com/reactioncommerce/reaction",
  "email": "engineering@reactioncommerce.com",
  "repository": {
    "type": "git",
    "url": "https://github.com/reactioncommerce/reaction.git",
    "directory": "packages/api-core"
  },
  "author": "Reaction Commerce <hello@reactioncommerce.com>",
  "license": "GPL-3.0",
  "bugs": {
    "url": "https://github.com/reactioncommerce/reaction/issues"
  },
  "scripts": {
    "test": "jest --maxWorkers=4",
    "test:unit:watch": "jest --maxWorkers=4 --watch",
    "test:unit:file": "jest --no-cache --maxWorkers=4",
    "test:unit:file:watch": "jest --no-cache --maxWorkers=4 --watch",
    "lint:gql": "graphql-schema-linter ./src/**/*.graphql"
  },
  "keywords": [],
  "graphql-schema-linter": {
    "rules": [
      "arguments-have-descriptions",
      "deprecations-have-a-reason",
      "descriptions-are-capitalized",
      "enum-values-have-descriptions",
      "enum-values-sorted-alphabetically",
      "fields-have-descriptions",
      "input-object-fields-sorted-alphabetically",
      "input-object-values-have-descriptions",
      "relay-connection-types-spec",
      "relay-page-info-spec",
      "type-fields-sorted-alphabetically",
      "types-are-capitalized",
      "types-have-descriptions"
    ]
  },
  "dependencies": {
    "@apollo/federation": "^0.38.1",
    "@apollo/server": "^4.2.1",
    "@graphql-tools/merge": "^8.3.12",
    "@graphql-tools/schema": "^9.0.10",
    "@reactioncommerce/api-utils": "^2.0.0",
    "@reactioncommerce/logger": "^1.1.3",
    "@reactioncommerce/random": "^1.0.2",
    "@reactioncommerce/reaction-error": "^1.0.1",
    "body-parser": "^1.19.0",
    "callsite": "^1.0.0",
    "cors": "^2.8.5",
    "cuid": "^2.1.8",
    "dataloader": "^2.0.0",
    "diehard": "^1.5.2",
    "dotenv": "^16.0.1",
    "envalid": "^6.0.1",
    "express": "^4.17.1",
    "graphql-iso-date": "^3.6.1",
    "graphql-redis-subscriptions": "^2.5.0",
    "graphql-subscriptions": "^2.0.0",
    "graphql-tag": "^2.10.0",
    "graphql-ws": "^5.11.2",
    "ioredis": "^5.2.4",
    "lodash": "^4.17.15",
    "mongodb": "4.4.1",
    "promise-retry": "^1.1.1",
    "simpl-schema": "^1.7.0",
    "ws": "^8.11.0"
  },
  "devDependencies": {
    "@babel/core": "~7.9.0",
    "@babel/preset-env": "~7.9.0",
    "@reactioncommerce/babel-remove-es-create-require": "~1.0.0",
    "@reactioncommerce/data-factory": "~1.0.1",
    "babel-plugin-rewire-exports": "~2.1.0",
    "babel-plugin-transform-es2015-modules-commonjs": "~6.26.2",
    "babel-plugin-transform-import-meta": "~1.0.1",
    "graphql": "^16.6.0"
  },
  "peerDependencies": {
    "graphql": "^16.6.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
