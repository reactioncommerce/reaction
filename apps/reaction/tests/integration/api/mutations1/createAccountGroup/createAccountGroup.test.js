import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import insertPrimaryShop from "@reactioncommerce/api-utils/tests/insertPrimaryShop.js";
import Factory from "/tests/util/factory.js";
import { importPluginsJSONFile, ReactionTestAPICore } from "@reactioncommerce/api-core";

const AddAccountToGroupMutation = importAsString("./CreateAccountGroupMutation.graphql");

jest.setTimeout(300000);

let createAccountGroup;
let customerGroup;
let mockAdminAccount;
let shopId;
let shopOpaqueId;
let testApp;

// Should match UTC datetime of the form YYYY-MM-DDTHH:MM:SS.mmmZ e.g. 2019-12-10T13:37:58.833Z
const UTC_REGEX_PATTERN = /\b[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}[.]{1}[0-9]{3}Z\b/;

beforeAll(async () => {
  testApp = new ReactionTestAPICore();
  const plugins = await importPluginsJSONFile("../../../../../plugins.json", (pluginList) => {
    // Remove the `files` plugin when testing. Avoids lots of errors.
    delete pluginList.files;

    return pluginList;
  });
  await testApp.reactionNodeApp.registerPlugins(plugins);
  await testApp.start();
  shopId = await insertPrimaryShop(testApp.context);

  const adminGroup = Factory.Group.makeOne({
    _id: "adminGroup",
    createdBy: null,
    name: "admin",
    permissions: ["reaction:legacy:groups/create"],
    slug: "admin",
    shopId
  });
  await testApp.collections.Groups.insertOne(adminGroup);

  mockAdminAccount = Factory.Account.makeOne({
    _id: "mockAdminAccount",
    groups: [adminGroup._id],
    shopId
  });
  await testApp.createUserAndAccount(mockAdminAccount);

  customerGroup = Factory.Group.makeOne({
    createdBy: null,
    name: "customer",
    permissions: [],
    slug: "customer",
    shopId
  });
  await testApp.collections.Groups.insertOne(customerGroup);

  shopOpaqueId = encodeOpaqueId("reaction/shop", shopId);

  createAccountGroup = testApp.mutate(AddAccountToGroupMutation);
});

// There is no need to delete any test data from collections because
// testApp.stop() will drop the entire test database. Each integration
// test file gets its own test database.
afterAll(() => testApp.stop());

test("anyone can add account to group if they have ALL the group permissions", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  const group = {
    description: "a group for testing purposes",
    name: "test-int-group",
    permissions: ["test-perm-1", "test-perm-2"],
    slug: "test-int-group"
  };

  const result = await createAccountGroup({ shopId: shopOpaqueId, group });

  expect(result.createAccountGroup.group).toMatchObject({
    _id: expect.any(String),
    createdBy: {
      _id: "cmVhY3Rpb24vYWNjb3VudDptb2NrQWRtaW5BY2NvdW50"
    },
    description: "a group for testing purposes",
    name: "test-int-group",
    permissions: ["test-perm-1", "test-perm-2"],
    shop: {
      _id: shopOpaqueId
    },
    slug: "test-int-group",
    updatedAt: expect.stringMatching(UTC_REGEX_PATTERN),
    createdAt: expect.stringMatching(UTC_REGEX_PATTERN)
  });
});


test("should throw if the group to be created has the same slug and shopId as an existing group", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  const group = {
    description: "a group for testing purposes",
    name: "test-int-group",
    permissions: ["test-perm-1", "test-perm-2"],
    slug: "test-int-group"
  };

  try {
    // Make sure DB is in clean state
    await testApp.collections.Groups.deleteMany({
      name: "test-int-group",
      slug: "test-int-group"
    });

    await createAccountGroup({ shopId: shopOpaqueId, group });

    // This should throw
    await createAccountGroup({ shopId: shopOpaqueId, group });
  } catch (errors) {
    expect(errors[0]).toMatchSnapshot();
  }

  const groups = await testApp.collections.Groups.find({
    name: "test-int-group",
    slug: "test-int-group"
  }).toArray();

  expect(groups.length).toBe(1);
});
