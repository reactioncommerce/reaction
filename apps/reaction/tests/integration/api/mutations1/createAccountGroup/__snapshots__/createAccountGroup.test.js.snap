// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw if the group to be created has the same slug and shopId as an existing group 1`] = `
Object {
  "extensions": Object {
    "code": "INTERNAL_SERVER_ERROR",
    "exception": Object {
      "details": Object {},
      "error": "conflict",
      "eventData": Object {},
      "isClientSafe": true,
      "reason": "Group already exist for this shop",
    },
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "Group already exist for this shop",
  "path": Array [
    "createAccountGroup",
  ],
}
`;
